### PR TITLE
Use primary transaction for useTransactionDisplayData

### PR DESCRIPTION
**Pull request by tmashuang**
Fallback to initialTransaction if primaryTransaction is not present.

This is suppose to mainly fix how we present the subsequent transaction of a failed transaction.
Since the failed transaction has the same nonce as the next proposed transaction, the data passed to the transaction by nonce will be the initial transaction data which is not true.

Ideally, we would always use the primaryTransaction data as the latest txCategory, value, from, to, etc.

Depends on #9040, since the txs will show primary state and the txCategory is not present.

### CODE REVIEW

File: ui/app/hooks/useTransactionDisplayData.js [hunk 1]
```diff
@@ -57,19 +57,19 @@ export function useTransactionDisplayData (transactionGroup) {
   const t = useI18nContext()
   const { initialTransaction, primaryTransaction } = transactionGroup
   // initialTransaction contains the data we need to derive the primary purpose of this transaction group
-  const { transactionCategory } = initialTransaction
+  const transactionCategory = primaryTransaction?.transactionCategory || initialTransaction?.transactionCategory
```
Comment (brad-decker) [inline ui/app/hooks/useTransactionDisplayData.js:60]: [Req] let's create a new variable above this that's just 'transaction' and set it to `primaryTransaction || initialTransaction` to reduce the complexity of the rest of the function.

**Review by brad-decker — CHANGES_REQUESTED** [review]

**Comment by brad-decker** [comment]
are there any types of transactions where we would expect the `initialTransaction` to be the correct thing to use? I modeled this after the original logic, which used `initialTransaction`, but if it always makes sense to prefer `primaryTransaction` do we need to fallback to initial? does primary not always get set?

**Comment by tmashuang** [comment]
> are there any types of transactions where we would expect the initialTransaction to be the correct thing to use?

I could be wrong, but I would say `primaryTransaction` data is what we use 99% of the time, expect for maybe the `initialTransaction` time that we use to determine when the tx was created.

**Comment by brad-decker** [comment]
>I could be wrong, but I would say primaryTransaction data is what we use 99% of the time, expect for maybe the initialTransaction time that we use to determine when the tx was created.

I'm still looking through all of this, but here's a link to the state of the transaction-list-item before the refactor. `initialTransaction` is passed to the component from the container as 'transaction'

`primaryTransaction` is used elsewhere in the container.

https://github.com/MetaMask/metamask-extension/blob/e06fb2c9f6612a4b4a324a8f09530e294eb43485/ui/app/components/app/transaction-list-item/transaction-list-item.container.js#L99

**Comment by brad-decker** [comment]
But primaryTransaction is passed in too, and is used here: :| 
https://github.com/MetaMask/metamask-extension/blob/e06fb2c9f6612a4b4a324a8f09530e294eb43485/ui/app/components/app/transaction-list-item/transaction-list-item.component.js#L142

**Review by brad-decker — DISMISSED** [review]

**Comment by brad-decker** [comment]
After reviewing the old code, I think that using the initialTransaction for all the displayData for a transaction was a mistake made during the refactor.

**Review by Gudahtt — CHANGES_REQUESTED** [review]
I don't think this was a mistake made during the transaction history rewrite; the use of the initial transaction was an intentional decision made to handle cancellations. If the user cancels a transaction, we still want the item in the transaction history to display the initial transaction details (category, amounts, etc.).

I think this bug dates back to the initial implementation of the "grouped" display of transactions in the UI, which was years ago.

Ensuring the transaction category is attached to speedups makes sense, but for cancellations this would still (I think?) result in us displaying the wrong thing, since the other details of the transaction (e.g. the tx data, the amount sent, etc.) no longer matches what the user actually tried to do. Transactions are cancelled by sending zero ETH to yourself, and users will likely be confused if their simple send transactions suddenly show as sending zero ETH to themselves after they cancel. We can't just use the initial transaction either, as then we'd fall victim to the same bug you're addressing here where the initial transaction is unrelated to the one currently being processed.

What we need here is something like... "the most recent transaction, excluding cancellations" 🤔 The selector doesn't return anything like that now, but it could.

**Comment by tmashuang** [comment]
I believe 913c237 handles the edge case for cancelled transactions without too much refactoring. We show the `initialTransaction` when the `primaryTransaction.type`  is `cancel`, but show the status of the `primaryTransaction`.

**Comment by brad-decker** [comment]
> I believe [913c237](https://github.com/MetaMask/metamask-extension/commit/913c237e384b556e3c9e80c891af0b30aa76769d) handles the edge case for cancelled transactions without too much refactoring. We show the `initialTransaction` when the `primaryTransaction.type` is `cancel`, but show the status of the `primaryTransaction`.

If all we need to do is use the canceled status instead of the `initialTransaction`'s status, I think the changes to useTransactionDisplayData need to be dropped. in the latest commit, the code has still been changed to prefer `primaryTransaction` over `initialTransaction` for every status except cancel.

File: ui/app/hooks/useTransactionDisplayData.js [hunk 2]
```diff
@@ -56,20 +56,21 @@ export function useTransactionDisplayData (transactionGroup) {
   const knownTokens = useSelector(getTokens)
   const t = useI18nContext()
   const { initialTransaction, primaryTransaction } = transactionGroup
-  // initialTransaction contains the data we need to derive the primary purpose of this transaction group
-  const { transactionCategory } = initialTransaction
+  const transaction = isCancelledTransaction(primaryTransaction) ? initialTransaction : primaryTransaction || initialTransaction
```
Comment (brad-decker) [inline ui/app/hooks/useTransactionDisplayData.js:59]: I think we should probably undo this change and simply override the status variable to use the canceled status when the `primaryTransaction` is canceled?

**Review by brad-decker — CHANGES_REQUESTED** [review]

File: ui/app/hooks/useTransactionDisplayData.js (hunk 2 above)
Comment (tmashuang) [reply ui/app/hooks/useTransactionDisplayData.js:59]: But this recommendation will still result in showing the wrong data that we want since `primaryTransaction.txParams` is different from `initialTransaction.txParams`. And we sort of do this already in https://github.com/MetaMask/metamask-extension/blob/develop/ui/app/helpers/utils/transactions.util.js#L225-L227

Comment (brad-decker) [reply ui/app/hooks/useTransactionDisplayData.js:59]: You're right. I think I am confusing myself sifting through the comments so far and the code. I'm gonna take another look at what you have.

Comment (brad-decker) [inline ui/app/hooks/useTransactionDisplayData.js:59]: Last question I have for both you and @Gudahtt -- in what cases would the `initialTransaction` be set and the `primaryTransaction` not be set? Just the first transaction in a group?

Comment (tmashuang) [reply ui/app/hooks/useTransactionDisplayData.js:59]: I think we always set the primaryTransactions? https://github.com/MetaMask/metamask-extension/blob/develop/ui/app/selectors/transactions.js#L253-L261

**Comment by tmashuang** [comment]
looking into a bug that gets introduced by this PR where Pending Transactions aren't transitioning to a complete state after some time.

**Comment by Gudahtt** [comment]
>We show the `initialTransaction` when the `primaryTransaction.type is cancel`, but show the status of the `primaryTransaction`.

Wouldn't this still show the wrong thing if the initial transaction was an unrelated failed tranasction? e.g. what if the user does the following:

1. The user attempts a ssimple send transaction with a nonce of 5
2. Transaction is rejected by the network for some reason (e.g. gas price too low)
3. Attempt to send a token, nonce of 5 is used (since the earlier tx with a nonce of 5 is no longer pending)
4. Transaction is pending for a while, because the gas price is low
5. The user attempts to cancel the this pending transaction

At that point we don't want to use the initial transaction, because then it would display as a simple send. We don't want to use the cancellation either, because that also would be a simple send. We need another reference, to the most-recent-excluding-cancellations transaction.

**Comment by tmashuang** [comment]
> We don't want to use the cancellation either, because that also would be a simple send

I believe this to be incorrect. In the background tx list, these would be two different transactions.

**Comment by tmashuang** [comment]
In c8c9eea, I have opted to specifically target the case that was presented to me where the failed transactions, which never gets submitted to the network and doesn't generate a tx hash. This will not work with failed transactions with hashes. For instance a tx that has a low gas price, waiting to get mined, but will eventually fail due to the gas being used up.

**Comment by tmashuang** [comment]
Going to need to reproduce #9435 with the branch, since it is an edge case that is going to effect how we present transaction data.

**Comment by brad-decker** [comment]
>We need another reference, to the most-recent-excluding-cancellations transaction, to correctly display this as a token send (or we need to group things differently)

Why isn't this synonymous with `primaryTransaction`? I'm reading through the `nonceSortedTransactions` selector code to really become familiar with it. What caught my eye is this:

```js
 } else if (nonce in nonceToTransactionsMap) {
        const nonceProps = nonceToTransactionsMap[nonce]
        insertTransactionByTime(nonceProps.transactions, transaction)

        if (status in PRIORITY_STATUS_HASH) {
          const { primaryTransaction: { time: primaryTxTime = 0 } = {} } = nonceProps

          if (status === CONFIRMED_STATUS || txTime > primaryTxTime) {
            nonceProps.primaryTransaction = transaction
          }
        }
```

We only update the `primaryTransaction` if `status` is in `PRIORITY_STATUS_HASH`. Wouldn't this preclude a cancellation from becoming the `primaryTransaction`? This check doesn't happen if `nonce` is not a key in `nonceToTransactionsMap` so it seems to me that the only way for a cancellation to be the `primaryTransaction` is if we process it first in this selector and it is the most recent transaction in terms of time. 

Am I offbase? @Gudahtt @tmashuang

**Comment by brad-decker** [comment]
The original issue this was targeting was resolved in #9687, I'm not sure this change makes sense with that done. Closing, but if you feel any piece of this still makes sense let me know!
